pkg/routeagent_driver/handlers/ovn/submarinerGWRouteHandler.go

```diff
@@ -0,0 +1,148 @@
+		smClient:        smClientSet,
+		remoteEndpoints: map[string]*submarinerv1.Endpoint{},
+	}
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) Init() error {
+	logger.Infof("Starting GatewayRouteHandler")
+
+	nextHopIP, err := getNextHopOnK8sMgmtIntf()
+	if err != nil || nextHopIP == "" {
+		return errors.Wrapf(err, "error getting the ovn kubernetes management interface IP")
+	}
+
+	gatewayRouteHandler.nextHopIP = nextHopIP
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) GetName() string {
+	return "submariner-gw-route-handler"
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) GetNetworkPlugins() []string {
+	// TODO enable when we switch to new implementation
+	// return []string{cni.OVNKubernetes}
+	return []string{}
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) RemoteEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	gatewayRouteHandler.remoteEndpoints[endpoint.Name] = endpoint
+
+	if gatewayRouteHandler.isGateway {
+		_, err := gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Create(context.TODO(),
+			gatewayRouteHandler.newGatewayRoute(endpoint), metav1.CreateOptions{})
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "error processing the remote endpoint creation for %q", endpoint.Name)
+		}
+	}
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) RemoteEndpointRemoved(endpoint *submarinerv1.Endpoint) error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	delete(gatewayRouteHandler.remoteEndpoints, endpoint.Name)
+
+	if gatewayRouteHandler.isGateway {
+		if err := gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Delete(context.TODO(),
+			endpoint.Name, metav1.DeleteOptions{}); err != nil {
+			return errors.Wrapf(err, "error deleting gatewayRoute %q", endpoint.Name)
+		}
+	}
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) TransitionToNonGateway() error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	gatewayRouteHandler.isGateway = false
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) TransitionToGateway() error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	gatewayRouteHandler.isGateway = true
+
+	for _, endpoint := range gatewayRouteHandler.remoteEndpoints {
+		if _, err := gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Update(context.TODO(),
+			gatewayRouteHandler.newGatewayRoute(endpoint), metav1.UpdateOptions{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				_, err = gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Create(context.TODO(),
+					gatewayRouteHandler.newGatewayRoute(endpoint), metav1.CreateOptions{})
+				return errors.Wrapf(err, "error updating gatewayRoute %q", endpoint.Name)
+			}
+		}
```

Comment:
It seems you want admiral's `util.CreateOrUpdate` here.

Reply:
when I try to pass h.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace) util.CreateOrUpdate it is giving me error updateStatus is missing.